core/include/traccc/fitting/kalman_filter/kalman_actor.hpp

```diff
@@ -71,12 +77,107 @@ struct kalman_actor_state {
 
     /// @return true if the iterator reaches the end of vector
     TRACCC_HOST_DEVICE
-    bool is_complete() {
-        if (!backward_mode && m_it == m_track_states.end()) {
-            return true;
-        } else if (backward_mode && m_it_rev == m_track_states.rend()) {
-            return true;
+    bool is_complete() const {
+        return ((!backward_mode && m_it == m_track_states.end()) ||
+                (backward_mode && m_it_rev == m_track_states.rend()));
+    }
+
+    /// @returns the current number of holes in this state
+    TRACCC_HOST_DEVICE
+    unsigned int count_missed() const {
+        unsigned int n_missed{0u};
+
+        if (backward_mode) {
+            for (const auto& trk_state : m_track_states) {
+                if (trk_state.smoothed().is_invalid()) {
+                    ++n_missed;
+                }
+            }
+        } else {
+            for (const auto& trk_state : m_track_states) {
+                if (trk_state.filtered().is_invalid()) {
+                    ++n_missed;
+                }
+            }
```

Comment:
For the sake of preventing divergence (impact probably minor):

```suggestion
        for (const auto& trk_state : m_track_states) {
            bool valid = backward_mode ? trk_state.smoothed().is_invalid() : trk_state.filtered().is_invalid();
            if (valid) {
                ++n_missed;
            }
        }
```